modules/system-disaster-recovery/src/main/java/org/apache/ignite/internal/disaster/system/MetastorageRepairImpl.java

```diff
@@ -0,0 +1,181 @@
+
+import static java.util.stream.Collectors.toMap;
+import static java.util.stream.Collectors.toSet;
+import static org.apache.ignite.internal.util.CollectionUtils.difference;
+import static org.apache.ignite.internal.util.CompletableFutures.allOf;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.ignite.internal.cluster.management.ClusterManagementGroupManager;
+import org.apache.ignite.internal.cluster.management.topology.LogicalTopology;
+import org.apache.ignite.internal.cluster.management.topology.api.LogicalNode;
+import org.apache.ignite.internal.cluster.management.topology.api.LogicalTopologyEventListener;
+import org.apache.ignite.internal.disaster.system.message.BecomeMetastorageLeaderMessage;
+import org.apache.ignite.internal.disaster.system.message.MetastorageIndexTermRequestMessage;
+import org.apache.ignite.internal.disaster.system.message.MetastorageIndexTermResponseMessage;
+import org.apache.ignite.internal.disaster.system.message.SystemDisasterRecoveryMessagesFactory;
+import org.apache.ignite.internal.disaster.system.repair.MetastorageRepair;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.network.MessagingService;
+import org.apache.ignite.internal.raft.IndexWithTerm;
+import org.apache.ignite.network.ClusterNode;
+
+/**
+ * Implementation of {@link MetastorageRepair}.
+ */
+public class MetastorageRepairImpl implements MetastorageRepair {
+    private static final IgniteLogger LOG = Loggers.forClass(MetastorageRepairImpl.class);
+
+    private final MessagingService messagingService;
+    private final LogicalTopology logicalTopology;
+    private final ClusterManagementGroupManager cmgManager;
+
+    private final SystemDisasterRecoveryMessagesFactory messagesFactory = new SystemDisasterRecoveryMessagesFactory();
+
+    /** Constructor. */
+    public MetastorageRepairImpl(
+            MessagingService messagingService,
+            LogicalTopology logicalTopology,
+            ClusterManagementGroupManager cmgManager
+    ) {
+        this.messagingService = messagingService;
+        this.logicalTopology = logicalTopology;
+        this.cmgManager = cmgManager;
+    }
+
+    @Override
+    public CompletableFuture<Void> repair(Set<String> participatingNodeNames, int metastorageReplicationFactor) {
+        LOG.info("Starting MG repair [participatingNodes={}, replicationFactor={}].", participatingNodeNames, metastorageReplicationFactor);
+
+        return waitTillValidatedNodesContain(participatingNodeNames)
+                .thenCompose(unused -> collectMetastorageIndexes(participatingNodeNames))
+                .thenCompose(indexes -> {
+                    LOG.info("Collected metastorage indexes [indexes={}].", indexes);
```

Comment:
I think that 4 spaces of padding are missing here

Reply:
Lambdas only require 4 spaces, and this is what there is here. An example from `TransactionsExample`:

```
                    accounts
                        .getAsync(tx, key)
                        .thenCompose(account -> {
                            account.balance += 300.0d;
```